### ISSUE TITLE
(Better) Node.js builtin emulation for client-side JS

### ISSUE THREAD

**Issue by Leftium** [sveltejs/kit#8307]
### Describe the problem

I've found many NPM packages that almost work in SvelteKit client-side JS, but their access to node builtins like `process` and `buffer` results in errors. For example:

- `ReferenceError: process is not defined` ([json-schema-to-zod](https://www.npmjs.com/package/json-schema-to-zod))
- `ReferenceError: process is not defined` ([google-photos-album-image-url-fetch](https://www.npmjs.com/package/google-photos-album-image-url-fetch))
- `Module "path" has been externalized for browser compatibility. Cannot access "path.extname" in client code.` ([gphotos-scraper](https://www.npmjs.com/package/gphotos-scraper))

There is a work-around for these errors, but it's neither obvious nor simple:
- [Fix for local dev errors](https://medium.com/@ftaioli/using-node-js-builtin-modules-with-vite-6194737c2cd2).
- [Fix for build-time errors](https://stackoverflow.com/a/72773693/117030)

Most average Svelte devs may just assume it's not possible (like I did at first).

### Describe the proposed solution

Ideally SvelteKit would provide something similar to Parcel's [Polyfilling & Excluding Builtin Node Modules](https://parceljs.org/features/node-emulation/#polyfilling-%26-excluding-builtin-node-modules)

The SvelteKit DX would be greatly enhanced by integrating the work-arounds linked above as either default or optional configurations.

If the work-arounds are not included (by default), detecting this class of error and giving meaningful error messages pointing to the correct work-arounds would still be very helpful.

### Alternatives considered

- It is possible to use the work-arounds linked above.
- Limiting use of the problem modules to server-side JS, and wrapping with internal endpoints.

### Importance

would make my life easier

### Additional Information

Simple repro demonstrating problem:

- https://kit-bugs-git-node-builtins-bug-leftium.vercel.app/
- https://github.com/Leftium/kit-bugs/tree/node-builtins-bug

**Comment by bluwy** [comment]
I think the core problem stems from Vite's stance of not polyfilling nodejs modules or syntax by default. The downside of doing this automatically is that it bloats the browser bundle. I don't think SvelteKit should change this, instead it uncovering the problematic dependencies is a good way to push the ecosystem towards not using nodejs modules.

**Comment by Conduitry** [comment]
Is there a straightforward way in Vite of opting in to pollyfilling? Is there something we want to link to from the Kit site? A commonly used Vite plugin?

**Comment by madeleineostoja** [comment]
I would also vote for just a more helpful generic error message along the lines of `Sveltekit doesn't support Node.js APIs in the browser, use [X workaround](link)` instead of bundling polyfills automatically, which I think is an antipattern for clientside code.

**Comment by benmccann** [comment]
For reference, the actual offending library in the reproduction is `@apidevtools/json-schema-ref-parser`. which is used by `json-schema-to-zod` and it looks like it's probably already been fixed: https://github.com/APIDevTools/json-schema-ref-parser/pull/280. There's an issue that `json-schema-to-zod` is depending on an older deprecated version of this package, which fetches a pinned version of the package forcing you onto the old broken one. I sent a fix to `json-schema-to-zod` here: https://github.com/StefanTerdell/json-schema-to-zod/pull/33

**Comment by Leftium** [comment]
Just an update for some context:

I originally ran into this issue on projects that were still using older versions of SvelteKit, where it seemed to be much more of a problem.

The current SvelteKit API makes it much easier to to import these modules from from server-only code so the errors don't occur. 99% of the time, I think these troublesome modules can be called from server-only code like +page.server.js without these errors, then populate PageData (which is then passed to the client-side code.)

So perhaps that is what a more helpful error message could suggest. (Along with the polyfill work-arounds if the modules are truly needed on the client-side.)

**Comment by bluwy** [comment]
> Is there a straightforward way in Vite of opting in to pollyfilling? Is there something we want to link to from the Kit site? A commonly used Vite plugin?

There isn't otherwise everyone would opt-in to polyfiling and call it a day. Vite has some discussions that might help, but there doesn't seem to be one solid answer:

- https://github.com/vitejs/vite/discussions/8799
- https://github.com/vitejs/vite/discussions/7915
- https://github.com/vitejs/vite/discussions/2785

There is https://github.com/sodatea/vite-plugin-node-stdlib-browser which could work. Maybe a new FAQ entry would be nice.

**Comment by dominikg** [comment]
-1 to document polyfilling node only api to the browser as a valid option to work around badly built third party modules.

Documenting how to properly use serverside only code (which is done in a pretty consistent way already imho) and directing questions there should suffice. If at all we should add a warning that you should not polyfill node api to the browser except for very special usecases and if you have to wonder if yours one of the special ones, it very likely isn't.

**Comment by madeleineostoja** [comment]
> If at all we should add a warning that you should not polyfill node api to the browser except for very special usecases and if you have to wonder if yours is one of the special ones, it very likely isn't.

I hate to be That Guy but it’s this kind of gatekeepy elitism that turned me off svelte in the early days. Documentation shouldn’t be opinionated and preachy. While I agree shipping node polyfills to the browser isn’t a good solution in a perfect world, we don’t live in a perfect world and I can’t speak for use-cases that aren’t mine, and nor can you.

**Comment by Rich-Harris** [comment]
Yeah, documentation is the answer here — maybe an entry on the FAQ or a new 'Troubleshooting' question so that people can search phrases like 'process is not defined' or '__dirname is not defined'

**Comment by bluwy** [comment]
https://github.com/vitejs/vite/pull/11877 improves the error message a bit for `Module "path" has been externalized for browser compatibility...`, linking to Vite's troubleshooting docs. `process is not defined` is harder to catch as it happens anywhere in runtime. But with Vite's PR, I think it's enough and could consider this issue done.

**Comment by Leftium** [comment]
For reference, I just posted an [answer on StackOverflow for a related question](https://stackoverflow.com/a/75550951/117030). I think this text would be a good candidate for the documentation:

> Unfortunately, nodejs libraries (as opposed to browser or client-side libraries) like `ccxt` are [not officially supported in client-side SvelteKit][1]; they are only supported in server-side SvelteKit.
> 
> **Generally, the best solution is to limit usage of nodejs libraries to server-side SvelteKit.** This should be sufficient 99% of the time:
> 
> - Call the nodejs library from [+page.server.js][5] and return values from the load function. The values can be accessed from +page.svelte via the `data` prop.
> - OR call the nodejs library from an API endpoint in [+server.js][4]. Then `fetch()` the appropriate internal API endpoint URL from +page.svelte (or almost anywhere else).
> 
> 
> For the 1% of the time this is not sufficient, you will have to either polyfill nodejs built-in dependencies and/or remove nodejs dependencies from the library (`ccxt`) in question.
> 
> 
> ---
> 
> Sometimes polyfilling will get some nodejs libraries to work client-side. (Buffer is a nodejs built-in module):
> 
> 1. [To fix local dev errors][2].
> 2. [To fix build-time errors][3].






  [1]: https://github.com/sveltejs/kit/issues/8307
  [2]: https://stackoverflow.com/a/72773693/117030
  [3]: https://medium.com/@ftaioli/using-node-js-builtin-modules-with-vite-6194737c2cd2
  [4]: https://kit.svelte.dev/docs/routing#server
  [5]: https://kit.svelte.dev/docs/routing#page-page-server-js
